src/uct/cuda/cuda_ipc/cuda_ipc_md.c

```diff
@@ -142,43 +233,94 @@ static ucs_status_t uct_cuda_ipc_mem_dereg(uct_md_h md, uct_mem_h memh)
     return UCS_OK;
 }
 
+
+static void uct_cuda_ipc_md_close(uct_md_h uct_md)
+{
+    uct_cuda_ipc_md_t *md = ucs_derived_of(uct_md, uct_cuda_ipc_md_t);
+
+    if (md->uuid_map) {
+        ucs_free(md->uuid_map);
+    }
+
+    if (md->peer_accessible_cache) {
+        ucs_free(md->peer_accessible_cache);
+    }
+
+    ucs_free(md);
+}
+
 static ucs_status_t
 uct_cuda_ipc_md_open(uct_component_t *component, const char *md_name,
                      const uct_md_config_t *config, uct_md_h *md_p)
 {
     static uct_md_ops_t md_ops = {
-        .close              = (void*)ucs_empty_function,
+        .close              = uct_cuda_ipc_md_close,
         .query              = uct_cuda_ipc_md_query,
         .mkey_pack          = uct_cuda_ipc_mkey_pack,
         .mem_reg            = uct_cuda_ipc_mem_reg,
         .mem_dereg          = uct_cuda_ipc_mem_dereg,
         .detect_memory_type = ucs_empty_function_return_unsupported,
     };
-    static uct_md_t md = {
-        .ops          = &md_ops,
-        .component    = &uct_cuda_ipc_component
-    };
 
-    *md_p = &md;
+    int num_devices;
+    uct_cuda_ipc_md_t* md;
+    uct_cuda_ipc_component_t* com;
+
+    UCS_STATIC_ASSERT(sizeof(md->peer_accessible_cache[0]) == sizeof(char));
+    UCT_CUDA_IPC_DEVICE_GET_COUNT(num_devices);
+
+    md = ucs_calloc(1, sizeof(uct_cuda_ipc_md_t));
+    if (md == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    md->super.ops       = &md_ops;
+    md->super.component = &uct_cuda_ipc_component.super;
+
+    /* allocate uuid map and peer accessible cache */
+    md->uuid_map_size     = 0;
+    md->uuid_map_capacity = 16;
+    md->uuid_map          = ucs_malloc(md->uuid_map_capacity * sizeof(CUuuid),
+                                       "uct_cuda_ipc_uuid_map");
+    if (md->uuid_map == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    /* Initially support caching accessibility of up to 16 other peers */
+    md->peer_accessible_cache = ucs_malloc(num_devices * md->uuid_map_capacity,
+                                           "uct_cuda_ipc_peer_accessible_cache");
```

Comment:
is there a case where number of peers(devices) change dynamically?

Reply:
the number of devices would not change during a run of the application